### ISSUE TITLE
Is is possible to use expects on a class method? 

### ISSUE THREAD

**Issue by Bennet-Sunder** [freerange/mocha#487]
I have seen examples in the docs where instance methods are verified using `expects` like this https://mocha.jamesmead.org/Mocha/ObjectMethods.html. 
Is it possible to check if a class method was invoked as part of the test cases. 
Something like this 
```
class Tasks < ActiveRecord::Base
  def self.do_tasks
  end
end
```
and in the spec file, we can do something like this
`Tasks.expects(:do_tasks). at_least_once`

**Comment by floehopper** [comment]
@Bennet-Sunder 

Yes, it is - and the syntax is exactly as you have it!

I've added [a new issue](https://github.com/freerange/mocha/issues/488) to capture the idea of adding examples for this scenario.

I'm going to close this issue now on the assumption that it is resolved, but feel free to open it if not.

**Comment by Bennet-Sunder** [comment]
but the thing is it doesn't work for me when I do this. 
```
module User
  class Task < ActiveRecord::Base
    def self.do_tasks(&block)
    end
  end
end
```
and in the spec file

`User::Task.expects(:do_tasks).at_least_once`

Although when I debug the `do_tasks` method is being invoked.

**Comment by floehopper** [comment]
@Bennet-Sunder I saw a notification suggesting you're still having problems with this, but it looks like you've deleted the comment. Do let me know if you need help and I'll re-open the issue.

**Comment by Bennet-Sunder** [comment]
@floehopper nah it's cool now. I realized I had to set up the expectation before I invoked the method that called it :D

**Comment by Bennet-Sunder** [comment]
@floehopper One other thing is it possible to do expect a method being invoked with only certain parameters only once but allow other method signatures to be invoked as well. 
`Task.any_instance.expects(:add).with('initial').once`
but at the same time allow other parameters to `add` to be called without raising an exception?

**Comment by floehopper** [comment]
> @floehopper One other thing is it possible to do expect a method being invoked with only certain parameters only once but allow other method signatures to be invoked as well.
> `Task.any_instance.expects(:add).with('initial').once`
> but at the same time allow other parameters to `add` to be called without raising an exception?

You'd need to also explicitly stub the other scenarios, e.g. `Task.any_instance.stubs(:add).with(Not(equals('initial')))`.

**Comment by Bennet-Sunder** [comment]
> You'd need to also explicitly stub the other scenarios, e.g. Task.any_instance.stubs(:add).with(Not(equals('initial'))).

That does make sense but do you think there could be a need for just checking if the `expects` is achieved the expected number of times with the allowed parameters at the same time allowing others. 

`Task.any_instance.expects(:send_email).with(:users, instance_of(User)).once.allow_others`

Now, say when a task gets saved it will call send_email with different parameters which we do not have to explicitly stub but when we add `allow_others` to allow other parameters of this method which still expecting `with` for the expected number of time. 

Do you think this makes sense or is it far too specific to include?

**Comment by floehopper** [comment]
> Do you think this makes sense or is it far too specific to include?

It's certainly an interesting idea, although it is quite specific. I've created [a new issue](https://github.com/freerange/mocha/issues/489) to capture it and I'll give it some more thought.
